idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirReferenceShortener.kt

```diff
@@ -358,17 +418,47 @@ private class ElementsToShortenCollector(
         val callExpression = qualifiedCallExpression.selectorExpression as? KtCallExpression ?: return
 
         val calleeReference = functionCall.calleeReference
-        val callableId = findUnambiguousReferencedCallableId(calleeReference) ?: return
-
-        val scopes = shorteningContext.findScopesAtPosition(callExpression, namesToImport, towerContextProvider) ?: return
-        val availableCallables = shorteningContext.findFunctionsInScopes(scopes, callableId.callableName)
+        val calledSymbol = findUnambiguousReferencedCallableId(calleeReference) ?: return
+        processQualifiedAccess(calledSymbol, qualifiedCallExpression, callExpression, shorteningContext::findFunctionsInScopes)
+    }
 
+    private fun processQualifiedAccess(
+        calledSymbol: FirCallableSymbol<*>,
+        qualifiedCallExpression: KtDotQualifiedExpression,
+        expressionToGetScope: KtExpression,
+        findCallableInScopes: (List<FirScope>, Name) -> List<AvailableSymbol<FirCallableSymbol<*>>>,
+    ) {
+        val option = callableShortenOption(calledSymbol)
+        if (option == ShortenOption.DO_NOT_SHORTEN) return
+        val callableId = calledSymbol.callableId
+
+        val scopes = shorteningContext.findScopesAtPosition(expressionToGetScope, namesToImport, towerContextProvider) ?: return
+        val availableCallables = findCallableInScopes(scopes, callableId.callableName)
+
+        val nameToImport = if (calledSymbol is FirConstructorSymbol) {
+            // A constructor is imported by the class name.
+            calledSymbol.containingClass()?.classId?.asSingleFqName()
+        } else {
+            callableId.asImportableFqName(shorteningContext.firSession)
+        }
+        val (matchedCallables, otherCallables) = availableCallables.partition { it.symbol.callableId == callableId }
         val callToShorten = when {
-            availableCallables.isEmpty() -> {
-                val additionalImport = callableId.asImportableFqName()
-                additionalImport?.let { ShortenQualifier(qualifiedCallExpression, it) }
+            otherCallables.all { it.importKind == ImportKind.DEFAULT_ON_DEMAND } -> {
```

Comment:
So this check is on the more conservative side because we don't have the usage check yet. Let me add a TODO here as a reminder to change it to a more complete solution later.